apiserver/service/forward.go

```diff
@@ -0,0 +1,144 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package service
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/gin-gonic/gin"
+
+	"github.com/kubeagi/arcadia/apiserver/pkg/forwardrepo"
+)
+
+const (
+	queryModelID = "modelid"
+
+	// huggingface or modelscope
+	pathParamRepo = "repo"
+
+	queryParamRevision = "revision"
+
+	repoToken = "REPOTOKEN"
+
+	huggingfaceRepo = "huggingface"
+	modelscopeRepo  = "modelscope"
+)
+
+type (
+	FrowarAPI   struct{}
+	SummaryResp struct {
+		Summary string `json:"summary"`
+	}
+)
+
+func (f *FrowarAPI) Summary(ctx *gin.Context) {
+	modelID := ctx.Query(queryModelID)
+	revision := ctx.DefaultQuery(queryParamRevision, "main")
+	repo := ctx.Param(pathParamRepo)
+	token := ctx.GetHeader(repoToken)
+	if modelID == "" {
+		ctx.AbortWithStatusJSON(http.StatusBadRequest, gin.H{
+			"message": "The modelid parameter is required",
+		})
+		return
+	}
+
+	var (
+		summary string
+		err     error
+
+		instance forwardrepo.Forward
+	)
+
+	tp := &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+	}
+	switch repo {
+	case huggingfaceRepo:
+		instance = forwardrepo.NewHuggingFace(
+			forwardrepo.WithHFToken(token), forwardrepo.WithModelID(modelID), forwardrepo.WithRevision(revision), forwardrepo.WithTransport(tp),
+		)
+	case modelscopeRepo:
+		instance = forwardrepo.NewModelScope(
+			forwardrepo.WithHFToken(token), forwardrepo.WithModelID(modelID), forwardrepo.WithRevision(revision),
+		)
+	default:
+		ctx.AbortWithStatusJSON(http.StatusBadRequest, gin.H{
+			"message": fmt.Sprintf("Unsupported repository type %s", repo),
+		})
+		return
+	}
+	summary, err = instance.Summary(ctx.Request.Context())
+	if err != nil {
+		ctx.AbortWithStatusJSON(http.StatusInternalServerError, gin.H{
+			"message": fmt.Sprintf("failed to get model summary error: %s", err),
+		})
+		return
+	}
+	ctx.JSON(http.StatusOK, SummaryResp{
+		Summary: summary,
+	})
+}
+
+func (f *FrowarAPI) Revisions(ctx *gin.Context) {
+	modelID := ctx.Query(queryModelID)
+	repo := ctx.Param(pathParamRepo)
+	token := ctx.GetHeader(repoToken)
+	if modelID == "" {
+		ctx.AbortWithStatusJSON(http.StatusBadRequest, gin.H{
+			"message": "The modelid parameter is required",
+		})
+		return
+	}
+
+	var (
+		revisions forwardrepo.Revision
+		err       error
+		instance  forwardrepo.Forward
+	)
+
+	tp := &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+	}
+	switch repo {
+	case huggingfaceRepo:
+		instance = forwardrepo.NewHuggingFace(forwardrepo.WithHFToken(token), forwardrepo.WithModelID(modelID), forwardrepo.WithTransport(tp))
+	case modelscopeRepo:
+		instance = forwardrepo.NewModelScope(forwardrepo.WithHFToken(token), forwardrepo.WithModelID(modelID))
+	default:
+		ctx.AbortWithStatusJSON(http.StatusBadRequest, gin.H{
+			"message": fmt.Sprintf("Unsupported repository type %s", repo),
+		})
+		return
+	}
+	revisions, err = instance.Revisions(ctx.Request.Context())
```

Comment:
the code to get the instance is duplicated for Summary and Revisions, let's refine it.